proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,489 @@
+      override fun User.isValid(): Boolean {
+        return id > 0 && name.length > 0
+      }
+    }
+  }
+}
+```
+That'll be enough for resolving the extension.
+
+#### 3. Companion object for the constraint interface we're looking for
+
+In case there's neither evidence in the companion of the target type, we'll look in the companion of the constraint interface:
+
+```kotlin
+interface Validator<A> {
+  fun A.isValid(): Boolean
+
+  companion object {
+    extension class GroupValidator<A>(with val userValidator: Validator<User>) : Validator<Group> {
+      override fun Group.isValid(): Boolean {
+        for (x in users) {
+          if (!x.isValid()) return false
+        }
+        return true
+      }
+    }
+  }
+}
+```
+
+#### 4. Subpackages of the package where the target type is defined
+
+The next step would be to look into the subpackages of the package where the target type (`User`) is declared.
+
+```kotlin
+package com.domain.repository
+
+import com.domain.User
+
+extension object UserRepository: Repository<User> {
+  override fun loadAll(): List<User> {
+    return listOf(User(25, "Bob"))
+  }
+
+  override fun loadById(id: Int): User? {
+    return if (id == 25) {
+      User(25, "Bob")
+    } else {
+      null
+    }
+  }
+}
+```
+
+Here we got a `Repository<User>` defined in a subpackage of `com.domain`, where the `User` type is defined.
+
+#### 5. Subpackages of the package where the constraint interface is defined
+
+Last place to look at would be subpackages of the package where the constraint interface is defined.
+
+```kotlin
+package com.data.instances
+
+import com.data.Repository
+import com.domain.User
+
+extension object UserRepository: Repository<User> {
+  override fun loadAll(): List<User> {
+    return listOf(User(25, "Bob"))
+  }
+
+  override fun loadById(id: Int): User? {
+    return if (id == 25) {
+      User(25, "Bob")
+    } else {
+      null
+    }
+  }
+}
+```
+
+Here, we are resolving it from `com.data.instances`, which a subpackage of `com.data`, where our constraint `Repository` is defined.
+
+## Error reporting
+
+We've got a `CallChecker` in place to report inlined errors using the context trace. That allows us to report as many errors as possible in a single compiler pass. Also provide them in two different formats:
+
+#### Inline errors while coding (using inspections and red underline)
+
+Whenever you're coding the checker is running and proper unresolvable extension errors can be reported within IDEA inspections.
+
+![Screenshot 2019-04-05 at 12 51 11](https://user-images.githubusercontent.com/6547526/55622809-85580480-57a1-11e9-8254-0830d29593ec.png)
+
+#### Errors once you hit the "compile" button:
+
+Once you hit the "compile" button or run any compile command you'll also get those errors reported.
+
+![Screenshot 2019-04-05 at 12 59 54](https://user-images.githubusercontent.com/6547526/55623259-be44a900-57a2-11e9-927e-fe0150265ba5.png)
```

Comment:
Ideally all the screenshots remain in the context of the examples instead of mentioning things like `Semigroup`

Reply:
Fixed for both screenshots in 87eafe6.